Add plugin P089-DLbus, add new web page, add flashing input feature to P001

uwekaditz opened the pull request:
NEW: [P089] PlugIn P089-DLbus to read and decode the DL-Bus
The DL-Bus is used in heating control units e.g. sold by Technische Alternative (www.ta.co.at).
The idea for this plug-in is based on Martin Kropf's project UVR31_RF24 (https://github.com/martinkropf/UVR31_RF24)
The plug-in is tested and workis fine for the ESR21 device.
The plug-in should be also able to decode the information from the UVR31, UVR1611 and UVR 61-3 devices but not tested yet.
The selected input needs a voltage divider as follows because the DL-Bus runs on 12 volts.

CHG: [WebServer] Added a simplified page just showing the device and its value
Usefull for devices with limited screens (e.g. smartphones) to have an overview of all read values

CHG: [P001] Switch-Input Switch to detect flashing inputs within a given time period
Usefull to monitor a LED state of an external device that cannot be connected to the Wemos board electrically
For monitoring purpose a photo resistor with a pull-up resistor can be used.
The new logic for PLUGIN_001_BUTTON_TYPE_FLASHING_INPUT is:
	if there is a 'state' change
		-> output value will be set to 2
        if there is a following 'state' change within PCONFIG_FLOAT(2) ("Longpress min. interval (ms)")
        	-> output value will be kept at 2
	else
		-> output value will be set to the original input 'state'

uwekaditz committed:
P089 new DL-Bus

uwekaditz committed:
WebServer new page with device and values for limited screens (e.g. smartphones)

uwekaditz committed:
P001 new state 'flashing' to detect flasing inputs

TD-er commented:
Can you rename the plugin-ID to 92?

uwekaditz commented:
Hi Gijs,tomorrow  will rename the plug-in to ID092.I used Github- Desktop to check out the current project but I didn't see that ID089 is already in use.Best regards,UweVon meinem Samsung Galaxy S8 gesendet.
-------- Ursprüngliche Nachricht --------Von: Gijs Noorlander <notifications@github.com> Datum: 28.10.19  19:20  (GMT+01:00) An: letscontrolit/ESPEasy <ESPEasy@noreply.github.com> Cc: uwekaditz <mueller.dresden@t-online.de>, Author <author@noreply.github.com> Betreff: Re: [letscontrolit/ESPEasy] Add plugin P089-DLbus, add new web page, add flashing input feature to P001 (#2695) Can you rename the plugin-ID to 92?

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/letscontrolit/ESPEasy/pull/2695?email_source=notifications\u0026email_token=ANHQUKJKWKCAM2OQ2IE5TALQQ4UOZA5CNFSM4JF52RJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECN4VIA#issuecomment-547080864",
"url": "https://github.com/letscontrolit/ESPEasy/pull/2695?email_source=notifications\u0026email_token=ANHQUKJKWKCAM2OQ2IE5TALQQ4UOZA5CNFSM4JF52RJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECN4VIA#issuecomment-547080864",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

uzi18 commented on src/_P001_Switch.ino:
```diff
@@ -567,22 +597,77 @@ boolean Plugin_001(byte function, struct EventStruct *event, String& string)
 
               // reset Userdata so it displays the correct state value in the web page
-              UserVar[event->BaseVarIndex] = sendState ? 1 : 0;
+//              UserVar[event->BaseVarIndex] = sendState ? 1 : 0;
```
why did you disabled this?

uwekaditz replied:
With the flashing detection the state of the UserVar[event->BaseVarIndex] can be
0, 1 or 2 (flashing)!
A few lines above the commented line you can find the code
UserVar[event->BaseVarIndex] = output_value;

The commented line can be deleted, it was just a reminder for me.

uzi18 commented on src/_P001_Switch.ino:
```diff
@@ -655,14 +740,14 @@ boolean Plugin_001(byte function, struct EventStruct *event, String& string)
 
               // reset Userdata so it displays the correct state value in the web page
-              UserVar[event->BaseVarIndex] = sendState ? 1 : 0;
+//              UserVar[event->BaseVarIndex] = sendState ? 1 : 0;
```
and here?

uwekaditz replied:
With the flashing detection the state of the UserVar[event->BaseVarIndex] can be
0, 1 or 2 (flashing)!
A few lines above the commented line you can find the code
UserVar[event->BaseVarIndex] = output_value;

The commented line can be deleted, it was just a reminder for me.

giig1967g replied:
You can't remove that line.
Its meant for displaying the state in the pinstate page.

giig1967g replied:
And in the device page

uwekaditz replied:
The state is saved a few lines above, see attached picture
![Highlighted code](https://user-images.githubusercontent.com/55511593/67765790-3af3e900-fa4d-11e9-8b8a-253c60f152c7.JPG)

giig1967g replied:
I developed the logic one year ago, so I had to review it.

Let me explain the logic:
for example, the event=3 means double-click, but it's not a valid pin state.
Example: the device generates event=3 (double-click) and send the event to the Rules with the command "sendData(event)".
The UserVar is used to pass the event number to the Rules, but it's not necessarily a valid pinstate value (0 or 1 ONLY).
Therefore at the end of the procedure you should set the UserVar[event->BaseVarIndex] = the state of the pin (either 0 or 1).
It cannot be 3, otherwise you will see "3" in the pinstate page.

Is it clear?

uwekaditz committed:
Merge remote-tracking branch 'upstream/mega' into mega

uwekaditz committed:
PlugIn-ID for DLbus renamed to 092 WebServer_Datapage type of DeviceIndex changed from byte to deviceIndex_t

uwekaditz commented:
Hi,

 

I added the flashing state „2“ to show that the input is flashing with a certain frequency within the detection time.

That’s why the pinstate (UserVar[event->BaseVarIndex]) should be 0, 1 or 2.

I missed that the state “3” should not be shown, this is my mistake.

 

Best regards,

Uwe

 

Von: Plebs [mailto:notifications@github.com] 
Gesendet: Dienstag, 29. Oktober 2019 16:13
An: letscontrolit/ESPEasy
Cc: uwekaditz; Author
Betreff: Re: [letscontrolit/ESPEasy] Add plugin P089-DLbus, add new web page, add flashing input feature to P001 (#2695)

 

@giig1967g commented on this pull request.

  _____  

In src/_P001_Switch.ino <https://github.com/letscontrolit/ESPEasy/pull/2695#discussion_r340138732> :

>                  log += output_value;
                 addLog(LOG_LEVEL_INFO, log);
               }
               sendData(event);
 
               // reset Userdata so it displays the correct state value in the web page
-              UserVar[event->BaseVarIndex] = sendState ? 1 : 0;
+//              UserVar[event->BaseVarIndex] = sendState ? 1 : 0;

I developed the logic one year ago, so I had to review it.

Let me explain the logic:
for example, the event=3 means double-click, but it's not a valid pin state.
Example: the device generates event=3 (double-click) and send the event to the Rules with the command "sendData(event)".
The UserVar is used to pass the event number to the Rules, but it's not necessarily a valid pinstate value (0 or 1 ONLY).
Therefore at the end of the procedure you should set the UserVar[event->BaseVarIndex] = the state of the pin (either 0 or 1).
It cannot be 3, otherwise you will see "3" in the pinstate page.

Is it clear?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/letscontrolit/ESPEasy/pull/2695?email_source=notifications&email_token=ANHQUKMFBD463UOW7RDZ2ELQRBHHHA5CNFSM4JF52RJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCJSMDYA#discussion_r340138732> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/ANHQUKLBH4JTNY5WLWD7CF3QRBHHHANCNFSM4JF52RJA> .  <https://github.com/notifications/beacon/ANHQUKKV4YVYSMTJPUM6XTTQRBHHHA5CNFSM4JF52RJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCJSMDYA.gif>

giig1967g commented:
A pinstate is the current physical state of the pin: either 0 or 1 (or -1 if OFFLINE). It cannot be 2. Period.
It would confuse the current logic of the three switch plugins for input (GPIO, PCF and MCP).
2 is a logical state that indicates an interpretation of what the pin is receiving.

In my opinion you should create a new plugin to account for flashing input. Or you could modify the PULSE plugin that already exists.

Finally, the 001 plugin is already so complex that I would avoid addind new feature to it.

uwekaditz commented:
Hi,

 

I understand your negative attitude to change your plugin, it is already very complex as you mentioned.

 

I don’t want to modify the Plugin 003: Pulse either, because I don't have to count pulses, I just need the information that the GPIO is flashing with a certain frequency.

 

I’m going to create a new plugin detecting the logical state as well as a flashing of a GPIO.

 

Best regards,

Uwe

 

Von: Plebs [mailto:notifications@github.com] 
Gesendet: Dienstag, 29. Oktober 2019 17:06
An: letscontrolit/ESPEasy
Cc: uwekaditz; Author
Betreff: Re: [letscontrolit/ESPEasy] Add plugin P089-DLbus, add new web page, add flashing input feature to P001 (#2695)

 

A pinstate is the current physical state of the pin: either 0 or 1 (or -1 if OFFLINE). It cannot be 2. Period.
It would confuse the current logic of the three switch plugins for input (GPIO, PCF and MCP).
2 is a logical state that indicates an interpretation of what the pin is receiving.

In my opinion you should create a new plugin to account for flashing input. Or you could modify the PULSE plugin that already exists.

Finally, the 001 plugin is already so complex that I would avoid addind new feature to it.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/letscontrolit/ESPEasy/pull/2695?email_source=notifications&email_token=ANHQUKM5L3JUEMBF5Q6KK23QRBNNNA5CNFSM4JF52RJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECRCYTQ#issuecomment-547499086> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/ANHQUKNMYZG5QSNXECEOCYLQRBNNNANCNFSM4JF52RJA> .  <https://github.com/notifications/beacon/ANHQUKKRQDDCCMR6D6F77QLQRBNNNA5CNFSM4JF52RJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECRCYTQ.gif>

uwekaditz committed:
Update _P001_Switch.ino

The original P001_Switch.ino from Oct 23, 2019 is back

giig1967g commented:
Hi @uwekaditz 
it's not my plugin... I just modified it adding some features and modifying the pinstate page to display the pin statuses. :)
I am sorry if my answer seemed hard, it wasn't my intention.

uwekaditz commented:
Hi @giig1967g 
no problem, I changed P001 already back.
I thaught it would be easier to change an existing plugin for the flashing detection feature.
I have used the modified P001 already for more than one year (always updating it with the newest sources downloaded from GitHub) to detect the standby state of my dish washing machine (glued a photo resistor to the ON-LED) to switch it remotely ON if the sun is powering my solar modules.

giig1967g commented:
In any case, the ultimate decision on development is in charge of @TD-er .
So let's wait for his comment too.

uzi18 commented:
@uwekaditz maybe we can first ask @TD-er what he think about it?

TD-er commented:
I have not yet looked at the plugin changes and today was way too hectic for me. Tomorrow I am away for a big part of the day, so I cannot look into it before tomorrow evening.

uwekaditz committed:
Update _P036_FrameOLED.ino

[P036] NEW more OLED sizes (128x32, 64x48) added to the original 128x64 size
[P036] NEW Display button can be inverted
[P036] NEW Content of header is adjustable, also the alternating function
[P036] CHG Parameters sorted

TD-er commented on src/_P036_FrameOLED.ino:
```diff
@@ -298,6 +397,10 @@ boolean Plugin_036(byte function, struct EventStruct *event, String& string)
         if (UserVar[event->BaseVarIndex] == 1) {
           // Display is on.
+          OLEDIndex = PCONFIG(8);
+          HeaderContent = eHeaderContent(PCONFIG(9));
```
There are only 8 values for PCONFIG.
So you're altering settings of the "next" task.

TD-er commented:
Can you split the changes for P036_Framed.ino into a new PR?
Now I cannot merge this one if I wanted to.

uwekaditz commented:
Closed for Updating P036
